idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt

```diff
@@ -0,0 +1,132 @@
+
+            val fqName = expression.getCallableDescriptor()?.fqNameUnsafe?.asString() ?: return@expressionVisitor
+
+            if (!fqName.matches(REGEX_RANGE_TO) && !fqName.matches(REGEX_UNTIL)) return@expressionVisitor
+
+            when {
+                fqName.matches(REGEX_RANGE_TO) -> visitRangeToExpression(expression, holder)
+                fqName.matches(REGEX_UNTIL) -> visitUntilExpression(expression, holder)
+                else -> return@expressionVisitor
+            }
+
+        }
+    }
+
+    private fun visitRangeToExpression(expression: KtExpression, holder: ProblemsHolder) {
+        if (expression.getArguments()?.second?.deparenthesize()?.isSizeMinusOne() != true) return
+
+        val constantValue = expression.getArguments()?.first?.constantValueOrNull()
+        val rightValue = (constantValue?.value as? Number)?.toInt() ?: return
+        if (rightValue != 0) return
+
+
+        holder.registerProblem(
+            expression,
+            "'rangeTo or the '..' call should be replaced with .indices call",
+            ProblemHighlightType.GENERIC_ERROR_OR_WARNING,
+            ReplaceManualRangeWithIndicesCallQuickFix()
+        )
+
+    }
+
+    private fun visitUntilExpression(expression: KtExpression, holder: ProblemsHolder) {
+        if (expression.getArguments()?.second?.deparenthesize()?.isSizeOrLength() != true) return
+
+        val constantValue = expression.getArguments()?.first?.constantValueOrNull()
+        val rightValue = (constantValue?.value as? Number)?.toInt() ?: return
+        if (rightValue != 0) return
+
+
+        holder.registerProblem(
+            expression,
+            "'rangeTo or the '..' call should be replaced with .indices call",
+            ProblemHighlightType.GENERIC_ERROR_OR_WARNING,
+            ReplaceManualRangeWithIndicesCallQuickFix()
+        )
+
+    }
+
+
+    private fun KtExpression.isSizeMinusOne(): Boolean {
+        if (this !is KtBinaryExpression) return false
+        if (operationToken != KtTokens.MINUS) return false
+
+        val leftValue = left?.isSizeOrLength() ?: return false
+        val constantValue = right?.constantValueOrNull()
+        val rightValue = (constantValue?.value as? Number)?.toInt() ?: return false
+        return rightValue == 1 && leftValue
+    }
+
+}
+
+class ReplaceManualRangeWithIndicesCallQuickFix : LocalQuickFix {
+    override fun getName() = "Replace with indices"
+
+    override fun getFamilyName() = name
+
+    override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
+        val element = descriptor.psiElement as KtExpression
+        val args = element.getArguments() ?: return
+        val second = (args.second as? KtBinaryExpression) ?: return
+        val array = (second.left as? KtDotQualifiedExpression)?.getLeftMostReceiverExpression() ?: return
+
+        element.replace(
+            KtPsiFactory(element).createExpressionByPattern(
+                "$0.indices",
```

Comment:
I'm not sure about this-- we should only replace these calls if the collection receiver actually has `.indices`, and this has never been checked.
An additional check should be made whether or not this size or length call is done on [a type that supports indices](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/indices.html).